middle_end/flambda2/types/meet_and_join_new.ml

```diff
@@ -267,6 +267,120 @@ let combine_results2 env ~meet_a ~meet_b ~left_a ~right_a ~left_b ~right_b
   combine_results env ~meet_ops:[meet_a; meet_b] ~left_inputs:[left_a; left_b]
     ~right_inputs:[right_a; right_b] ~rebuild:(fun (a, (b, ())) -> rebuild a b)
 
+type ext =
+  | No_extensions
+  | Ext of
+      { when_a : TEE.t;
+        when_b : TEE.t
+      }
+
+let meet_disjunction ~meet_a ~meet_b ~bottom_a ~bottom_b ~meet_type
+    ~join_env_extension initial_env val_a1 val_b1 extensions1 val_a2 val_b2
+    extensions2 =
+  let join_scope = TE.current_scope initial_env in
+  let env = TE.increment_scope initial_env in
+  let to_extension scoped_env =
+    TE.cut scoped_env ~cut_after:join_scope
+    |> Typing_env_level.as_extension_without_bindings
+  in
+  let direct_return r =
+    map_env r ~f:(fun scoped_env ->
+        TE.add_env_extension_strict initial_env (to_extension scoped_env)
+          ~meet_type:(New meet_type))
+  in
+  let env_a, env_b = Or_bottom.Ok env, Or_bottom.Ok env in
+  let env_a, env_b =
+    match extensions1 with
+    | No_extensions -> env_a, env_b
+    | Ext { when_a; when_b } ->
+      ( Or_bottom.bind env_a ~f:(fun env ->
+            TE.add_env_extension_strict env when_a ~meet_type:(New meet_type)),
+        Or_bottom.bind env_b ~f:(fun env ->
+            TE.add_env_extension_strict env when_b ~meet_type:(New meet_type)) )
+  in
+  let env_a, env_b =
+    match extensions2 with
+    | No_extensions -> env_a, env_b
+    | Ext { when_a; when_b } ->
+      ( Or_bottom.bind env_a ~f:(fun env ->
+            TE.add_env_extension_strict env when_a ~meet_type:(New meet_type)),
+        Or_bottom.bind env_b ~f:(fun env ->
+            TE.add_env_extension_strict env when_b ~meet_type:(New meet_type)) )
+  in
+  let a_result : _ meet_result =
+    match env_a with
+    | Bottom -> Bottom (New_result ())
+    | Ok env -> meet_a env val_a1 val_a2
+  in
```

Comment:
Not new but writing it here just in case: if `env_a` ends up equal to `Bottom` because of the extensions (say, because there were extensions in `extensions1`'s `when_a`), we will have `a_result` equal to `Bottom (New_result ())` even if `val_a2` was already bottom, losing the `Right_input` information. We could perhaps improve this by passing additional `is_obvisously_bottom_{a,b}` functions, do you think this would be worth the change or not?

Reply:
I expect that when one side is obviously Bottom, there will be no associated extensions. We should probably turn this into an invariant check at some point.

Reply:
Ah yes, there should be no associated extensions coming from that side, and extensions coming from the other side should not be enough to cause a bottom, because they should have been turned into a bottom in that case (I think that's still possible but should be rare enough for it not to matter).